### PR TITLE
refactored MasterClass and EventBriteEvent

### PR DESCRIPTION
**Pull request by JonNorman**
### What does this change?

In anticipation of parsing another EventBrite feed (for LiveEvents) the `MasterClass` model and agents have been refactored to be offshoots of an `Event`.
### What is the value of this and can you measure success?

It will enable a smoother introduction of the live events component
### Does this affect other platforms - Amp, Apps, etc?

No

/cc @kelvin-chappell

### CODE REVIEW

File: commercial/app/model/commercial/events/MasterClassAgent.scala [hunk 1]
```diff
@@ -0,0 +1,95 @@
+package model.commercial.events
```
Comment (regiskuckaertz) [inline commercial/app/model/commercial/events/MasterClassAgent.scala]: minor detail: I think this file should be named `MasterclassAgent`

**Comment by regiskuckaertz** [comment]
Awesome! Can't wait to see how the event component comes to life. There are a few files still using the 'masterClasses' case that you might want to rename.

**Comment by JonNorman** [comment]
@regiskuckaertz Yep I'm on that - I entered into a *slight* world of pain before when I tried to address this - git doesn't recognise case-only changes by default, and altering this and making the changes caused a bunch of errors across my different branches and duplicated the directories - one for Masterclass and one for MasterClass :sob: Will try again!

File: commercial/app/model/commercial/events/Event.scala [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+      firstParagraph = event.firstParagraph,
+      keywordIdSuffixes = Nil,
+      mainPicture = None
+    )
+  }
+}
+
+trait TicketHandler{
+  def tickets: Seq[Ticket]
+  def status: String
```
Comment (kelvin-chappell) [inline commercial/app/model/commercial/events/Event.scala]: `status` probably doesn't belong here.  I think it's used to filter events out of the list.

File: commercial/app/model/commercial/events/Event.scala [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+    val priceList = tickets.map(_.price).sorted.distinct
+    if (priceList.size > 1) {
+      val (low, high) = (priceList.head, priceList.last)
+      f"£$low%,.2f to £$high%,.2f"
+    } else f"£${priceList.head}%,.2f"
+  }
+
+  lazy val ratioTicketsLeft = 1 - (tickets.map(_.quantitySold).sum.toDouble / tickets.map(_.quantityTotal).sum)
+
+  lazy val isOpen = { status == "live" }
```
Comment (kelvin-chappell) [inline commercial/app/model/commercial/events/Event.scala]: Same as for `status`.

File: commercial/app/model/commercial/events/Event.scala [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+import org.jsoup.nodes.{Document, Element}
+import play.api.libs.json._
+import play.api.libs.functional.syntax._
+
+case class Ticket(price: Double, quantityTotal: Int, quantitySold: Int, donationEvent: Boolean)
+
+object Ticket{
+
+  def buildTicket(hidden: Boolean, donation: Boolean, valuePence: Double, quantityTotal: Int, quantitySold: Int): Option[Ticket] = {
+    if (!hidden) {
```
Comment (kelvin-chappell) [inline commercial/app/model/commercial/events/Event.scala]: Logic would be better the other way round.

File: commercial/app/model/commercial/events/Event.scala [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+                 status: String,
+                 venue: Venue,
+                 tickets: Seq[Ticket],
+                 capacity: Int,
+                 firstParagraph: String,
+                 tags: Seq[String] = Nil) extends TicketHandler
+
+object Event{
+
+  private def extractFirstParagraph(html: String) = {
```
Comment (kelvin-chappell) [inline commercial/app/model/commercial/events/Event.scala]: This is only used by `apply` so could be moved in there.

File: commercial/app/model/commercial/events/Event.scala [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+
+  lazy val truncatedFirstParagraph = StringUtils.abbreviate(firstParagraph, 250)
+}
+
+object Masterclass {
+  private val guardianUrlLinkText = "Full course and returns information on the Masterclasses website"
+
+  def apply(event: Event): Masterclass = {
+
+    def extractGuardianUrl: String = {
```
Comment (kelvin-chappell) [inline commercial/app/model/commercial/events/Event.scala]: This is really an `Option[String]`

File: commercial/app/views/masterClasses/masterClasses.scala.html [hunk 7]
```diff
@@ -22,22 +22,22 @@ <h3 class="commercial__title">
             </div>
             <div class="commercial__body">
                 <ul class="lineitems">
-                    @events.take(4).map { event =>
+                    @masterclasses.take(4).map { masterclass =>
```
Comment (kelvin-chappell) [inline commercial/app/views/masterClasses/masterClasses.scala.html]: Probably better to `take 4` in the controller and just pass the list to the template.

**Comment by kelvin-chappell** [comment]
Looks good!  Just some minor comments.

File: commercial/app/model/commercial/events/Eventbrite.scala [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+        response.events.filter(_.startDate.isAfter(now.plusWeeks(2)))
+      }
+
+      def parsePagesOfEvents(feedMetaData: FeedMetaData, feedContent: => Option[String]): Future[ParsedFeed[EBEvent]] = {
+
+        feedMetaData.switch.isGuaranteedSwitchedOn flatMap { switchedOn =>
+          if (switchedOn) {
+            val start = currentTimeMillis
+            feedContent map { body =>
+              val JsArray(pages) = Json.parse(body).as[JsArray]
```
Comment (kelvin-chappell) [inline commercial/app/model/commercial/events/Eventbrite.scala]: This could be done in one go as
`val responses = Json.parse(body).as[Seq[EBResponse]]`

**Comment by kelvin-chappell** [comment]
Just a very small suggestion above, and then :+1: :ship:

**Pull request merged**